# QBIO 401 HW 1
## Hirad Hosseini
### QBIO 401, Fall 2023

Write a Python function that takes as input a sequence string and returns a list with 4
entries that are the number of A, C, G, and T in the sequence.

In [1]:
sample_seq = "ACTGACTGACTG"
def count_bases(seq):
    ACTG_count = [0,0,0,0]
    for base in seq:
        if base == "A":
            ACTG_count[0] += 1
        elif base == "C":
            ACTG_count[1] += 1
        elif base == "T":
            ACTG_count[2] += 1
        elif base == "G":
            ACTG_count[3] += 1
    return ACTG_count
count_bases(sample_seq)

[3, 3, 3, 3]

Write another Python function that takes two inputs: a sequence string and a string of
two letters (e.g., “CG” or “CT”). This function returns the number of times the two
letters occur consecutively in the sequence.

In [12]:
sample_motif = "CT"
def count_motif(seq,motif):
    motif_count = 0
    for base in range(len(seq) - 1):
        if seq[base] == motif[0] and seq[base + 1] == motif[1]:
            motif_count += 1
    return motif_count
count_motif(sample_seq, sample_motif)

3

Write a third Python function that takes as input a sequence string and returns a list
with 16 entries that are the outputs of function #2 for all 16 possible two letter strings.

In [14]:
'''

Our motif_counts list is ordered such that we permutate through all doublets starting with A, then doublets
starting with C, then with T, and finally with G. Thus, our list is ordered such that it represents the
counts for [AA, AC, AT, AG, CA, CC, CT, CG, TA, ...]

'''

def motif_permutator(seq):
    motif_counts = [0] * 16
    bases = ["A", "C", "T", "G"]
    i = 0
    for base1 in bases:
        for base2 in bases:
            motif = [base1, base2]
            motif_counts[i] = count_motif(seq, motif)
            i += 1
    return motif_counts
motif_permutator(sample_seq)

[0, 3, 0, 0, 0, 0, 3, 0, 0, 0, 0, 3, 2, 0, 0, 0]

For each of the two FASTA files, has the output for functions #1 and #3. (There are many
possible output formats, list them in alphabetical order, i.e., AA, AC, AG, AT, CA, etc.)

In [15]:
#From lecture2functions.py:

def loadFASTA2(filename):
    '''Outputs a sequence string from the FASTA file named filename'''
    seqlist = []
    with open(filename,"r") as infile:
      temp = infile.readline() # reads a single line
                               # we don't want the first line 
                               # of the FASTA file ">..."
      for line in infile: # reads one line at a time
                          # when it gets to the end of the file it stops
        temp = line.replace("\n","") # removes \n
        temp = temp.replace("\r","") # removes \r too
        seqlist.append(temp)
    seq = "".join(seqlist) # combines the list into a string
    return seq

In [20]:
#For human sequence of PTPN11
human_seq = loadFASTA2("/Users/hiradh/QBIO 401/human_sequence.fasta")

In [21]:
#Output list is ordered [A,C,T,G]
count_bases(human_seq)

[1773, 1139, 1751, 1410]

In [29]:
'''

Our output list is ordered such that we permutate through all doublets starting with A, then doublets
starting with C, then with T, and finally with G. Thus, our list is ordered such that it represents the
counts for [AA, AC, AT, AG, CA, CC, CT, CG, TA, ...]

'''

motif_permutator(human_seq)

[595, 280, 438, 459, 394, 275, 371, 99, 340, 308, 606, 497, 443, 276, 336, 355]

In [24]:
#For drosophila orthologue csw
drosophila_seq = loadFASTA2("../QBIO 401/drosophila_sequence.fasta")

In [26]:
count_bases(drosophila_seq)

[2395, 1876, 1718, 1675]

In [28]:
'''

Our output list is ordered such that we permutate through all doublets starting with A, then doublets
starting with C, then with T, and finally with G. Thus, our list is ordered such that it represents the
counts for [AA, AC, AT, AG, CA, CC, CT, CG, TA, ...]

'''

motif_permutator(drosophila_seq)

[870,
 521,
 523,
 481,
 621,
 452,
 348,
 455,
 404,
 402,
 495,
 416,
 499,
 501,
 352,
 323]

In a “markdown” cell answer the following question:
Since humans have much higher rates of methylation than Drosophila, we would expect
to see far fewer CpGs in humans (we will discuss this topic soon in lecture). Is this what
we see?

Based on our outputs for the drosophila and human sequences, it seems that this hypothesis holds. Based on the ordering system for our motif_permutator function output, the eight index of each list represents the count for the "CG" motif. The human sequence contains 99 instances of "CG" compared to 455 instances for the drosophila sequence. 